src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala

```diff
@@ -6,68 +6,128 @@ import is.hail.utils._
 import org.apache.spark.Partitioner
 
 class OrderedRVDPartitioner(
-  val numPartitions: Int,
   val partitionKey: Array[String], val kType: TStruct,
-  // rangeBounds is partition max, sorted ascending
-  // rangeBounds: Array[pkType]
+  // rangeBounds: Array[Interval[pkType]]
+  // rangeBounds is interval containing all partition keys within a partition
   val rangeBounds: UnsafeIndexedSeq) extends Partitioner {
-  require((numPartitions == 0 && rangeBounds.isEmpty) || numPartitions == rangeBounds.length + 1,
-    s"nPartitions = $numPartitions, ranges = ${ rangeBounds.length }")
+  val numPartitions: Int = rangeBounds.length
 
   val (pkType, _) = kType.select(partitionKey)
+  val pkIntType = TInterval(pkType)
+  val rangeBoundsType = TArray(pkIntType)
 
-  val pkKFieldIdx: Array[Int] = partitionKey.map(n => kType.fieldIdx(n))
-  val pkKOrd: UnsafeOrdering = OrderedRVDType.selectUnsafeOrdering(pkType, (0 until pkType.size).toArray, kType, pkKFieldIdx)
-
-  val rangeBoundsType = TArray(pkType)
   assert(rangeBoundsType.typeCheck(rangeBounds))
 
+  require(rangeBounds.isEmpty || rangeBounds.zip(rangeBounds.tail).forall { case (left: Interval, right: Interval) =>
+    !left.overlaps(pkType.ordering, right) && pkType.ordering.compare(left.start, right.start) <= 0
+  })
+
+  val rangeTree: IntervalTree[Int] = new IntervalTree[Int](IntervalTree.fromSorted(pkType.ordering,
+    Array.tabulate[(BaseInterval, Int)](numPartitions) { i =>
+      (rangeBounds(i).asInstanceOf[Interval], i)
+    }, 0, rangeBounds.size))
+
+  val pkKFieldIdx: Array[Int] = partitionKey.map(n => kType.fieldIdx(n))
+  val pkKOrd: ExtendedOrdering = OrderedRVDType.selectExtendedOrdering(pkType, (0 until pkType.size).toArray, kType, pkKFieldIdx)
```

Comment:
This is used in places where you can just extract the pk from the k as an annotation.  That's a useful function if we don't already have it.